### PR TITLE
Make Node coloring based on process status and allowed types manually settable

### LINKED ISSUE #82631: Nodes with disabled processing appear unclickable when selecting exported node
### Godot version

4.2 dev5

### System information

Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1060 (NVIDIA; 30.0.15.1403) - Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz (8 Threads)

### Issue description

image
Control can be selected, but doesn't look like.

### Steps to reproduce

1. ```
extends Node
@export var test: Control
```
2. Add a Control to the scene
3. Set its `process_mode` to disabled

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by Stronkkey**
Fixes #82631
Supersedes #82672

### CODE REVIEW

**Comment by KoBeWi** [comment]
When creating a property track in AnimationPlayer, all nodes are grayed out.
The NodePath bug I mentioned last time still exists and looks the same as Animation bug, so maybe they are related.

**Comment by Stronkkey** [comment]
Whoops didn't test that out properly.

**Comment by KoBeWi** [comment]
The logic of your setter methods looks wrong.
```C++
void SceneTreeEditor::set_node_item_process_color(bool p_enabled, TreeItem *p_item, Node *p_node) {
	if (!p_node->can_process() && p_enabled) {
```
Here, the node becomes only disabled when `p_enabled` is true. Which means that calling `set_node_type_color_enabled()` has no effect if you pass false.

```C++
valid = p_enabled;
```
This line in `set_node_item_type_color()` makes the nodes always disabled when calling `set_node_type_color_enabled()` with false. It's probably what originally caused the all-gray bug.

The code works correctly, but the implementation is quite messy[.](https://cdn.discordapp.com/attachments/841444502212182016/1174020943471972482/FB_IMG_1699978774983.jpg)
Though it's not like the code of SceneTreeEditor is super clean on itself. I think the setters you added should just enable the flag and don't do anything else, as they are only called when initializing the dialog. `_add_nodes()` already handles the color logic.

File: editor/editor_properties.cpp [hunk 1]
```diff
@@ -2810,7 +2810,11 @@ void EditorPropertyNodePath::_node_assign() {
 	if (!scene_tree) {
 		scene_tree = memnew(SceneTreeDialog);
 		scene_tree->get_scene_tree()->set_show_enabled_subscene(true);
-		scene_tree->set_valid_types(valid_types);
+		if (!valid_types.is_empty()) {
+			scene_tree->set_valid_types(valid_types);
+			scene_tree->get_scene_tree()->set_node_type_color_enabled(true);
+		}
```
Comment (KoBeWi) [inline editor/editor_properties.cpp:2816]: This change is wholly unnecessary I'd say. Instead you can make `set_node_item_type_color()` return early if valid types are empty.

File: editor/gui/scene_tree_editor.cpp [hunk 2]
```diff
@@ -209,6 +209,7 @@ void SceneTreeEditor::_add_nodes(Node *p_node, TreeItem *p_parent) {
 	}
 
 	TreeItem *item = tree->create_item(p_parent);
+	node_list.insert(item, p_node);
```
Comment (KoBeWi) [inline editor/gui/scene_tree_editor.cpp:212]: You also need to clear the list when Tree is cleared.

Comment (Stronkkey) [reply editor/gui/scene_tree_editor.cpp:212]: Should clearing the list in _update_tree be enough?

Comment (KoBeWi) [reply editor/gui/scene_tree_editor.cpp:212]: Yes, that the only place where `tree` is cleared.

**Review by KoBeWi — APPROVED** [review]
I am not entirely sure if keeping a HashMap of nodes is really necessary (the colors are usually enabled before any node is added, no?), but overall looks ok.

**Comment by Stronkkey** [comment]
That's a relic from the old design so I removed it.

**Comment by akien-mga** [comment]
There seems to be merge conflicts, would you be able to rebase the branch on latest `master`?

See [PR workflow](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase) if you're not familiar with rebasing.

**Pull request closed without merging**